filecoin-proofs/src/bin/phase2.rs

```diff
@@ -1,75 +1,71 @@
-/// A CLI program for running Phase2 of Filecoin's trusted-setup.
-///
-/// # Build
-///
-/// From the directory `rust-fil-proofs` run:
-///
-/// ```
-/// $ RUSTFLAGS="-C target-cpu=native" cargo build --release -p filecoin-proofs --bin=phase2
-/// ```
-///
-/// # Usage
-///
-/// ```
-/// # Create initial params for a circuit using:
-/// $ RUST_BACKTRACE=1 ./target/release/phase2 new \
-///     <--porep, --epost, --fpost> \
-///     [--poseidon (default), --sha-pedersen] \
-///     <--2kib, --8mib, --512mib, --32gib, --64gib>
-///
-/// # Contribute randomness to the phase2 params for a circuit:
-/// $ RUST_BACKTRACE=1 ./target/release/phase2 contribute <path to params file>
-///
-/// # Verify the transition from one phase2 params file to another:
-/// $ RUST_BACKTRACE=1 ./target/release/phase2 verify \
-///     --paths=<comma separated list of file paths to params> \
-///     --contributions=<comma separated list of contribution digests>
-///
-/// # Run verification as a daemon - verify the parameters and contributions as they are written to
-/// # the `rust-fil-proofs` directory:
-/// $ RUST_BACKTRACE=1 ./target/release/phase2 verifyd
-/// ```
-use std::fmt::{self, Display, Formatter};
 use std::fs::{self, File};
-use std::io::{BufReader, BufWriter};
+use std::io::{self, BufReader, BufWriter};
 use std::path::Path;
 use std::process::Command;
 use std::str::{self, FromStr};
-use std::thread::sleep;
+use std::sync::mpsc::{channel, TryRecvError};
+use std::thread::{self, JoinHandle};
 use std::time::{Duration, Instant};
 
 use clap::{App, AppSettings, Arg, ArgGroup, SubCommand};
+
 use filecoin_proofs::constants::*;
 use filecoin_proofs::parameters::{
     setup_params, window_post_public_params, winning_post_public_params,
 };
-use filecoin_proofs::types::*;
+use filecoin_proofs::types::{
+    PaddedBytesAmount, PoRepConfig, PoRepProofPartitions, PoStConfig, PoStType, SectorSize,
+};
 use filecoin_proofs::with_shape;
-use log::info;
-use phase2::{verify_contribution, MPCParameters};
-use rand::SeedableRng;
+use log::{error, info};
+use phase2::small::{read_small_params_from_large_file, MPCSmall, Streamer};
+use phase2::MPCParameters;
+use rand::rngs::OsRng;
+use rand::{RngCore, SeedableRng};
+use rand_chacha::ChaChaRng;
 use simplelog::{self, CombinedLogger, LevelFilter, TermLogger, TerminalMode, WriteLogger};
 use storage_proofs::compound_proof::{self, CompoundProof};
 use storage_proofs::hasher::Sha256Hasher;
 use storage_proofs::merkle::MerkleTreeTrait;
 use storage_proofs::porep::stacked::{StackedCircuit, StackedCompound, StackedDrg};
 use storage_proofs::post::fallback::{FallbackPoSt, FallbackPoStCircuit, FallbackPoStCompound};
 
+const CHUNK_SIZE: usize = 10_000;
```

Comment:
Was this value chosen because it's roughly the read buffer size divided by the byte length of a serialized G1Affine?

Reply:
The buffer size used by the `Streamer` is chosen so one chunk will fit into the buffer. This`CHUNK_SIZE` was chosen to be 'good enough': it keeps memory usage low without an observable performance penalty. It's conceivable that a different strategy for all of this might provide some micro-optimization, but this is what emerged heuristically and has been tested.